File: CIVITASRUBY/lib/juego_texto/vista_textual.rb
#encoding:utf-8
require_relative '../operaciones_juego'
require_relative '../respuestas'
require_relative '../salidas_carcel'
require_relative '../gestiones_inmobiliarias'
require_relative '../civitas_juego'
require 'io/console'

module Civitas

  class Vista_textual
    RESPUESTAS =[Respuestas::NO,Respuestas::SI]
    SALIDAS = [Salidas_carcel::PAGANDO,Salidas_carcel::TIRANDO]
    GESTIONES = [Gestiones_Inmobiliarias::VENDER,Gestiones_Inmobiliarias::HIPOTECAR,Gestiones_Inmobiliarias::CANCELAR_HIPOTECA,Gestiones_Inmobiliarias::CONSTRUIR_CASA,Gestiones_Inmobiliarias::CONSTRUIR_HOTEL,Gestiones_Inmobiliarias::TERMINAR]

    def initialize
      @juego_model = nil
      @i_gestion = -1
      @i_propiedad = -1
    end
    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end

    def salir_carcel
      opcion = menu("Elige la forma para intentar salir de la carcel",["Pagando","Tirando el dado"])
      return SALIDAS[opcion]
    end
    def comprar
      array = ["NO","SI"]
      opcion = menu("¿Deseas comprar la calle?", array)
      return RESPUESTAS[opcion]
    end

    def gestionar
      opcion = menu("Elige un tipo de gestión",["VENDER", "HIPOTECAR", "CANCELAR_HIPOTECA",
            "CONSTRUIR_CASA", "CONSTRUIR_HOTEL", "TERMINAR"])
      @i_gestion = opcion
      if GESTIONES[opcion] != Gestiones_Inmobiliarias::TERMINAR
          opcion = menu("Elige una propiedad a gestionar: ", @juego_model.get_propiedades_UI());
          @i_propiedad = opcion
      end
      
    end

    def get_gestion
      @i_gestion
    end

    def get_propiedad
      @i_propiedad
    end

    def mostrar_siguiente_operacion(operacion)
      puts "La siguiente opeacion que se va a realizar es " + operacion.to_s
    end

    def mostrar_eventos
      diario = Diario.instance
      while diario.eventos_pendientes == true
        puts diario.leer_evento + "\n"
      end
    end

    def set_civitas_juego(civitas)
         @juego_model=civitas
         self.actualizar_vista
    end

    def actualizar_vista
      puts "\nJugador Actual : \n" + @juego_model.get_jugador_Actual().to_s + "\nCasilla Actual: \n" + @juego_model.get_casilla_actual().to_s
    end

    
  end

end

File: CIVITAS/src/JuegoTexto/VistaTextual.java
package JuegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.GestionesInmobiliarias;
import civitas.Jugador;
import civitas.TituloPropiedad;
import GUI.Respuestas;

class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() {
    int opcion = menu ("¿Deseas comprar la calle?",
        new ArrayList<> (Arrays.asList("SI","NO")));    

    return (Respuestas.values()[opcion]);
  }

  void gestionar () {
    int opcion = menu ("Elige un tipo de gestión",
        new ArrayList<> (Arrays.asList("VENDER", "HIPOTECAR", "CANCELAR_HIPOTECA",
            "CONSTRUIR_CASA", "CONSTRUIR_HOTEL", "TERMINAR" ))); 
    this.iGestion = opcion;
    
    
    if(GestionesInmobiliarias.values()[opcion] != GestionesInmobiliarias.TERMINAR)
     opcion = menu("Elige una propiedad a gestionar: ", this.juegoModel.getPropiedades_UI());
    
    this.iPropiedad = opcion;
             
  

  }
  
  public int getGestion(){return this.iGestion;}
  
  public int getPropiedad(){return this.iPropiedad;}
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      System.out.println("La siguiente opeacion que se va a realizar es " + operacion) ;
  }


  void mostrarEventos() {
     while(Diario.getInstance().eventosPendientes()){
         System.out.println(Diario.getInstance().leerEvento() + "\n");
     }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();

    }
  
  void actualizarVista(){
      System.out.println("\nJugador Actual : \n" + this.juegoModel.getJugadorActual().toString() 
              + "\nCasilla Actual: \n" +this.juegoModel.getCasillaActual().toString());
     
  } 
}
